pkg/validation/admission.go

```diff
@@ -0,0 +1,70 @@
+package validation
+
+import (
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/pkg/apis/metal3/v1alpha1"
+)
+
+// ValidateHostForAdmission takes a list of existing hosts, and a host we are
+// trying to add to the cluster.  It returns an error if the host fails
+// uniqueness validation, and cannot be added to the cluster.  The optional
+// third argument `old` denotes the previous version of `host` in the case of an
+// update.
+//
+// This is designed to be run by an admission webhook, `ValidateCreate`, and
+// `ValidateUpdate`.
+func ValidateHostForAdmission(existingHosts []metal3v1alpha1.BareMetalHost, host metal3v1alpha1.BareMetalHost, old *metal3v1alpha1.BareMetalHost) error {
+	if host.Spec.BMC.Address == "" && host.Spec.BootMACAddress == "" && host.Spec.BMC.CredentialsName == "" {
+		return nil
+	}
+
+	if len(existingHosts) == 0 {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+
+	for _, existingHost := range existingHosts {
+		if old != nil && old.Namespace == existingHost.Namespace && old.Name == existingHost.Name {
+			continue
+		}
+
+		// The BMC address is optional so only validate it if it's set
+		if host.Spec.BMC.Address != "" && host.Spec.BMC.Address == existingHost.Spec.BMC.Address {
```

Comment:
I think we might want to dig down a bit further rather than just comparing strings. e.g. redfish://192.168.122.1 and redfish-ilo5://192.168.122.1 are the same BMC.
I'm not sure how the path components come into it. Is it possible with e.g. redfish that multiple hosts share a BMC with the same IP address but different paths?

Reply:
Yes, chassis can have multiple hosts with 1 controller (think blade servers). Each host would have a different path, with (for example) the `/Systems/1` part changing based on which host is being used.

Reply:
Makes sense, so we should compare the hostname, path (for redfish), and presumably port (for IPMI/iDRAC/iRMC), but not the scheme.

Reply:
How much knowledge of that sort of thing do we want to embed here, vs. using the `AccessDetails` implementations to do the comparison?